File: 3.cpp
//
// Created by linkinpark213 on 5/6/19.
//
#include <iostream>

using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int max_length = 0;
        string substr;
        for (int i = 0; i < s.size(); i++) {
            substr = "";
            for (int j = i; j < s.size(); j++) {
                int pos = substr.find(s[j]);
                if (pos >= 0 && pos < j - i) {
                    break;
                }
                substr.push_back(s[j]);
                max_length = max(max_length, j - i + 1);
            }
        }
        return max_length;
    }
};

int main() {
    Solution solution;
    cout << solution.lengthOfLongestSubstring("abcabcbb");
}

File: pysrc/3.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) == 0:
            return 0
        l, r = 0, 0
        d = {}
        ans = 1
        while r < len(s):
            if s[r] not in d:
                d[s[r]] = r
                ans = max(ans, r - l + 1)
            else:
                while l < d[s[r]]:
                    d.pop(s[l])
                    l += 1
                l = d[s[r]] + 1
                d[s[r]] = r
            r += 1
        return ans


if __name__ == '__main__':
    solution = Solution()
    print(solution.lengthOfLongestSubstring("abcabcbb"))
    print(solution.lengthOfLongestSubstring("bbbbb"))
    print(solution.lengthOfLongestSubstring("pwwkew"))
